Notification script changes

Armored-Dragon opened the pull request:
Currently we have several "notification" scripts which are used individually for selective purposes. This PR should make it so that there is a single "notification" script that only handles system notifications, and armored_chat provides its own notification window.

I have moved the existing notification scripts to script-archive.

## Features
This PR adds the following:
- New UI for notifications
- New UI for chat messages
- Notification history
- Notification 'details' section. (You can read more about a notification if extra information was provided)

## TODO
   - [ ] Test pal notifications
   - [ ] Reimpliment chat notification sounds
   - [ ] Finish notification history UI
   - [ ] Snapshots use old notification system? 

## Examples
Notifications show in the top-right corner. The white bar at the bottom is a timeout indicator that slides from right to left.
![image](https://github.com/user-attachments/assets/e16d8d5c-cac7-4b37-9da7-fb13cd2385c6)

You can select and browse though your history of notifications in the event you missed one you might have needed.
This design is still work in progress.
![image](https://github.com/user-attachments/assets/b3f0232b-20aa-46b7-8d63-ecce32afb3b2)

Details expanded.
![image](https://github.com/user-attachments/assets/5deb4ede-6cc7-47c7-aac8-6c88937deefb)

This provides a new chat notification because notification-core has been moved.
![image](https://github.com/user-attachments/assets/ffa88304-d2aa-481a-b3cb-73e1825087cf)

AleziaKurdis commented:
I think your approach is risky... if you try to change everywhere how notification are going to be called, this is a good way to left a disaster behind... some are use only in apps and other scripts.

Also
when you say: "3d effect would be very nice." 
hmm. it not only to be nice, it won't be functional as a notifications if you don't. (who need a notification that we miss 50% of the time because it is on the hud and not where we are looking.)

If you need help to figure how those entrypoint can be called, just let me know, I went through this before.

Also, in VR I do want to be able to flush them without have to click. Mainly for the snapshot notification (it is why I added the "whoosh" move)

AleziaKurdis commented:
_"These never seem to work on my machine but it's there incase it does elsewhere?"_

Well, it works fine on mine.


I don't like too much the idea to go from an API method specific for notification to a generic communication method. How poeple will figure how to make a notification?



_"This could be a just me issue though. I know pal.js does not work if defaultScripts.js isn't also running. I don't normally have defaultScripts running."_

How can you not using that? Shouldn't we develop based on what will be deployed on the user machine ? Are you sure you have the right experience about this system?

Armored-Dragon commented:
I tested with a clean install of Overte and the `Window.` notifications did not work on a fresh install. I also test when I have defaultScripts loaded. 

> I don't like too much the idea to go from an API method specific for notification to a generic communication method. How poeple will figure how to make a notification?

This PR is still a work im progress and will be expanded as more work is being done. I only have a generic one presently and I only have an outline for the structure of connection based notifications.

Armored-Dragon commented:
I just reinstalled Overte on Windows and I found that the notifications do in fact still work as expected. Somehow running Overte on my Arch Linux install does not allow the notifications to function.

AleziaKurdis commented:
Is "Window.displayAnnouncement" works on linux ?

Armored-Dragon commented:
> to me
Window.notifyEditError
Window.notify
 should be replaced by
Window.displayAnnouncement
which is going by the API.

I agree. I changed everything in edit.js to use displayAnnouncement. I also added a 'details' field for that function, so if further details are needed for that announcement it can be provided.

Armored-Dragon commented on scripts/system/notifications/notifications.js:
```diff
@@ -0,0 +1,248 @@
+		if (sound) playSound.system();
+
+	},
+	connection: (text = "") => {
```
Yes. I am wanting to have a separate notification handling for connection based notifications in the future. Having this properly separated would require me changing the Pal app. I feel that would be a little bit too far out of the scope of this pull request so I left this code here to work on in the near future.
I have another pr with a pal app that is still work in progress.  https://github.com/overte-org/overte/pull/1422

Armored-Dragon commented on scripts/system/notifications/sound/systemNotification.mp3:
This does not sound like an error notification to me. We can probably find a new sound together sometime.

Armored-Dragon commented:
![image](https://github.com/user-attachments/assets/e7462ecd-646e-4516-b5ab-bcbe806e03d5)
![image](https://github.com/user-attachments/assets/a1a68535-ac91-4447-aaea-d20c5339e9c9)![image](https://github.com/user-attachments/assets/ac6d90b8-dc04-4030-9e6b-37897e03619b)

Armored-Dragon commented:
> NotificationsDesktop.qml is printing a debug message with JSON for every notification.

Fixed.

> All notifications have the ping noise,

This was a bug, which is now fixed.

> the ping noise sometimes gets cut off?

I did notice that too at one point and [I thought I had fixed that](https://github.com/overte-org/overte/pull/1633/commits/fa67e1ed89472df566d3bfa7300b922d2c2299d6).

> The notifications icon shaking is distracting for me.
> Can we avoid shaking icon by default?

Shaking is now entirely disabled. I left the code in so when https://github.com/overte-org/overte/issues/1691 gets addressed I can enable it again based on user settings.

> Could the chat notification get a slightly larger font,

I do not think the font can get much bigger than what it is now before it starts to look worse. Would you be interested in a accessibility option (similar to #1691 ) for people who are hard of seeing / have lower resolution monitors?

>  The darker description text box is also visible in the log even on notifications that don't use it

Fixed.

<img width="347" height="91" alt="image" src="https://github.com/user-attachments/assets/a4cdfedb-1fb8-4965-a2d4-1d75f8f0623a" />


> Can we make the un-hovered notification background a little darker and more opaque?

[Opacity was entirely removed](https://github.com/overte-org/overte/pull/1633/commits/4a3813ed239fbc6d14e54877f943c7895dd594ff). All transparency is now handled by the colors. The colors are also now less transparent too.

<img width="347" height="150" alt="image" src="https://github.com/user-attachments/assets/04ac32fb-d76f-4d1f-a91d-3100add65921" />


> Scrolling after hitting the ends of the notification log zooms the camera in desktop mode

I think that is an issue with how the desktop handles QML that is attached under it. If so that would be out of scope of this PR and would be a separate fix.

> There also seems to be one too many scroll "ticks", which might be a bug in the Flickable element. Scrolling all the way to the bottom, and scrolling up by one "tick", it scrolls up enough to fill the empty space at the bottom, but doesn't actually scroll up meaningfully until the next "tick".

Unless I misunderstand you, this issue happens at any point in the flickable element. Scroll once or twice down, and if you scroll back up, it just gets nudged. It might be in the same boat as the above issue. If you hover your mouse cursor over the scroll bar, everything will work as expected.

> "Do not disturb" label is slightly too high

I'm now forcing it to be aligned vertically. I hope this fixes the issue?

> The notification log is saved, but there isn't a clear button.

I disabled saving (for the time being) as I don't think there is a use of saving these notifications yet. There is now a clear button.

<img width="347" height="91" alt="image" src="https://github.com/user-attachments/assets/b392b774-d385-4f59-80fc-bafdb19b5981" />

Armored-Dragon commented:
> Scrolling the notification log now only does one scroll tick until it gets clicked again. That issue with the sticky top and bottom seems to be gone now though.

#1546 

> Uncaught SyntaxErro

Fixed.

> Is it intended that the notification bubbles only have a countdown bar in VR?

No, but it is available for me on desktop. I wouldn't know why you don't see the timeout bar in desktop.

JulianGro commented:
Some things I noticed while testing:
- The notification icon that is in the top right at all times has a different background and scale compared to the ones on the top left. I think it would look better if they were the same size and had the same background. Personally, I prefer the size and background of the already existing icons on the top left. The icons on the top left are also slightly transparent, which would be cool to have to the top right icon as well; That way it isn't as distracting.
The actual notifications themselves look fine the way they are. Maybe for the top right icon we could use the same scale, background, and transparency as the icons on the top left, but have it take on the same transparency and background as the notifications when there is a notification in the top right?
- Personally, I think it would look nicer to have the notifications show up as expanded by default.
- Clicking a notification doesn't expand it, but opens the list of notifications instead. Expanding works fine in said list.
- Non-expanded notifications have different scaling. Personally, the scaling of the expanded notifications looks nicer to me. 
<img width="372" height="196" alt="overte-snap-by--on-2025-08-13_11-13-19" src="https://github.com/user-attachments/assets/46a2b32a-04c4-4241-a9d0-dea44cff607f" />
<img width="371" height="183" alt="overte-snap-by--on-2025-08-13_11-13-41" src="https://github.com/user-attachments/assets/5bfa8f68-ff89-4a33-81b3-ab203c2613d5" />

- The indentation on chat message notifications only works on the first line.
<img width="645" height="160" alt="overte-snap-by--on-2025-08-13_11-15-09" src="https://github.com/user-attachments/assets/b8284cbb-0a45-4227-9a71-879b0b68e031" />

- Short chat notifications look pretty empty. Maybe increasing the font size for short messages would make it look a little less empty?
Instead, we could also decrease the size of the notification box. To keep it noticeable the notification box could be aligned to the right, so it is a little closer to the center of the screen.
<img width="647" height="97" alt="overte-snap-by--on-2025-08-13_11-14-40" src="https://github.com/user-attachments/assets/7f05b0b8-83f3-4773-8ec7-ef23ac907e92" />

- It would be cool if there was a way to dismiss notifications without deleting all of them. Maybe just right-clicking on them could delete them? And/or since the space to the right of the notification text is usually empty, maybe a little cross button for closing them would be nice.
- The "Do not disturb" toggle uses a different Switch design than the graphics settings. If they goal is to make a "unified" UI then they should use exactly the same Switch element, in my opinion. (The scale also doesn't match, but that seems less arbitrary since they will basically never both be on the same plane in VR.)
<img width="893" height="803" alt="Auswahl_059" src="https://github.com/user-attachments/assets/ec378336-cf8e-4598-95bc-85a11ecef55c" />

- Personally, I don't think the red trash can icon for the notifications looks very good when it is relatively small (on my 27inch 1440p screen for example). It's too small to have the lid look like anything more than noise. The trash can icon from the Create app would probably be a better fit, since its features are more well defined.
Generally, red on black background is a bit hard to see (especially on my Valve Index). Maybe doing something similar to the Create app, where it is a white trash can on a red background would look nicer? Something like a transparent slightly red tinted button with a strongly red tinted outline.

Armored-Dragon commented:
> The notification icon that is in the top right at all times has a different background and scale compared to the ones on the top left.

I am assuming that you are talking about the microphone and shield icons? I chose against adding the notification container there for a few reasons. It would be hard to have the notification history/list there as it is movable, and I have had a few instances of that container just disappearing. I was going to try and address that at some point in the future probably with #1052.
There is a obvious background around it because I find it very hard to see white icons on a lot of bright surfaces.
<img width="100" height="46" alt="image" src="https://github.com/user-attachments/assets/b9cf6aea-8efe-4257-8648-8dc67aee476d" />
Here both icons blend into the background.

> The icons on the top left are also slightly transparent, which would be cool to have to the top right icon as well; That way it isn't as distracting.

I tried doing that at the beginning, but ran into the same color issues I described above. Having the icons transparent made seeing them hard.

> Personally, I think it would look nicer to have the notifications show up as expanded by default.

I think having just a brief description of an announcement is fine. If there is confusion as to what an announcement means, or someone wants extra information about that, they should be responsible for getting that information. If the announcement isn't important to the user, it will take up a small amount of their screen space and disappear after the timeout.

> Clicking a notification doesn't expand it, but opens the list of notifications instead. Expanding works fine in said list.

That's intended. Do you want me to change it so that selecting the notification takes you immediately to that expanded listing?

> The indentation on chat message notifications only works on the first line.

That *was* intended, kind of. I agree that looks silly. I changed that with https://github.com/overte-org/overte/pull/1633/commits/f994f83b89b01672166cca45e71151b6a0a1f631.

> Short chat notifications look pretty empty.

Adjusted. The message bubbles now fit into the space with a cap. https://github.com/overte-org/overte/pull/1633/commits/67e8ac286473132361816e2cfeb9a7b8b56eb5fc.

<img width="612" height="246" alt="2" src="https://github.com/user-attachments/assets/8102b499-3ae4-41b5-b8ac-f5a157dc865d" />

> It would be cool if there was a way to dismiss notifications without deleting all of them.

That notification history is intended to be the entire history for a session. You are not really meant to "dismiss" a notification from there.

> The "Do not disturb" toggle uses a different Switch design than the graphics settings. 

You're right. The short of it is that I intended on completing the Settings UI, which would permit me to break out the internal components that the Settings UI uses so that it can be easily referenced in other apps. This did not pan out.

You may have noticed I started shipping a file named "Colors.qml" in some of my other PRs. The intent was to, in the near future, make a PR that rips out all of them, and make a single file that is referenced across all of the apps. The issue I had was that if I had pushed a Colors.qml in one PR, I wouldn't really be able to use that in another PR without causing issues. 

I made a quick change that adjusted the color of the slider. https://github.com/overte-org/overte/pull/1633/commits/22afcb9f2e436634e4b14e724c6e8ee802b1a140

> Personally, I don't think the red trash can icon for the notifications looks very good when it is relatively small

I'm now using the HiFi icon and I made it slightly less saturated (more white). https://github.com/overte-org/overte/pull/1633/commits/40d17765a6cd76d6ee5fc6f97e897affcd6e6924
<img width="306" height="50" alt="image" src="https://github.com/user-attachments/assets/fc1a66ff-12a7-4242-a55d-e881e135db0a" />
